File: activations/src/activations.cc
#include <cmath>
#include <activations/activations.h>
#include <swiss/strings.h>

namespace cerebrum {
  Activations::Activations(
      std::string& activation,
      swiss::WeightsXf input_weights) {
    swiss::to_lower(activation);

    // Get the activation by name
    auto activation_ = functions.at(activation);

    // Execute on the input layer
    activation_(input_weights);
  }

  // TODO(jparr721) - Docs here
  void Activations::ReLu(swiss::WeightsXf& input_weights) {
    for (size_t i = 0; i < input_weights.size(); ++i) {
      float val = input_weights(i);
      input_weights(i) = std::max(val, 0.0f);
    }
  }

  // TODO(jparr721) - Docs here
  void Activations::Sigmoid(swiss::WeightsXf& input_weights) {
    for (size_t i = 0; i < input_weights.size(); ++i) {
      input_weights(i) = 1 / 1 + std::exp(input_weights(i));
    }
  }

  // TODO(jparr721) - Docs here
  void Activations::Tanh(swiss::WeightsXf& input_weights) {
    for (size_t i = 0; i < input_weights.size(); ++i) {
      input_weights(i) = 2 / (1 + std::exp(-2 * input_weights(i))) - 1;
    }
  }

  void Activations::dReLu(swiss::WeightsXf& inputs) {

  }
} // namespace cerebrum

File: src/activations.rs
use std::f64;

enum ActivationKind {
    ReLu,
    Sigmoid,
    Tanh,
}

fn ReLu(z: f64) -> f64 {
    (0.0_f64).max(z)
}

fn Sigmoid(z: f64) -> f64 {
    (1.0 / 1.0 + -z.exp())
}

fn Tanh(z: f64) -> f64 {
    (1.0 - z.tanh().powf(2.0))
}

pub fn make_activation(activation: Option<ActivationKind>) -> Option<fn(f64) -> f64> {
    match ActivationKind {
        ActivationKind::ReLu => Some(ReLu),
        ActivationKind::Sigmoid => Some(Sigmoid),
        ActivationKind::Tanh => Some(Tang),
        _ => None,
    }
}
